File: platform/xdebugger-impl/src/com/intellij/xdebugger/impl/actions/handlers/XDebuggerStepIntoHandler.java
// Copyright 2000-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package com.intellij.xdebugger.impl.actions.handlers;

import com.intellij.openapi.actionSystem.DataContext;
import com.intellij.xdebugger.XDebugSession;
import com.intellij.xdebugger.XSourcePosition;
import com.intellij.xdebugger.impl.actions.XDebuggerSuspendedActionHandler;
import com.intellij.xdebugger.stepping.ForceSmartStepIntoSource;
import com.intellij.xdebugger.stepping.XSmartStepIntoHandler;
import com.intellij.xdebugger.stepping.XSmartStepIntoVariant;
import org.jetbrains.annotations.ApiStatus;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.concurrency.Promise;

import java.util.List;

@ApiStatus.Internal
public class XDebuggerStepIntoHandler extends XDebuggerSmartStepIntoHandler {
  @Override
  protected boolean isEnabled(@NotNull XDebugSession session, @NotNull DataContext dataContext) {
    return XDebuggerSuspendedActionHandler.isEnabled(session);
  }

  @Override
  protected <V extends XSmartStepIntoVariant> Promise<List<V>> computeVariants(XSmartStepIntoHandler<V> handler, XSourcePosition position) {
    return handler.computeStepIntoVariants(position);
  }

  @Override
  protected <V extends XSmartStepIntoVariant> boolean handleSimpleCases(XSmartStepIntoHandler<V> handler,
                                                                        List<? extends V> variants,
                                                                        XDebugSession session) {
    if (variants.size() == 1) {
      V singleVariant = variants.get(0);
      if (singleVariant instanceof ForceSmartStepIntoSource forceSmartStepIntoSource && forceSmartStepIntoSource.needForceSmartStepInto()) {
        return super.handleSimpleCases(handler, variants, session);
      }
    }

    if (variants.size() < 2) {
      session.stepInto();
      return true;
    }
    return false;
  }
}

File: platform/xdebugger-impl/src/com/intellij/xdebugger/impl/actions/handlers/XDebuggerStepIntoHandler.kt
// Copyright 2000-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package com.intellij.xdebugger.impl.actions.handlers

import com.intellij.openapi.actionSystem.DataContext
import com.intellij.xdebugger.impl.actions.XDebuggerProxySuspendedActionHandler
import com.intellij.xdebugger.impl.frame.XDebugSessionProxy
import com.intellij.xdebugger.impl.rpc.XDebugSessionApi
import com.intellij.xdebugger.impl.rpc.XSmartStepIntoTargetDto

internal class XDebuggerStepIntoHandler : XDebuggerSmartStepIntoHandler() {
  override fun isEnabled(session: XDebugSessionProxy, dataContext: DataContext): Boolean {
    return XDebuggerProxySuspendedActionHandler.isEnabled(session)
  }

  override suspend fun computeTargets(session: XDebugSessionProxy): List<XSmartStepIntoTargetDto> {
    return XDebugSessionApi.getInstance().computeStepTargets(session.id)
  }

  override suspend fun handleSimpleCases(targets: List<XSmartStepIntoTarget>, session: XDebugSessionProxy): Boolean {
    if (targets.size == 1) {
      val singleTarget = targets[0]
      if (singleTarget.needsForcedSmartStepInto) {
        return super.handleSimpleCases(targets, session)
      }
    }
    if (targets.size < 2) {
      XDebugSessionApi.getInstance().stepInto(session.id)
      return true
    }
    return false
  }
}
